### ISSUE TITLE
[0.2.38] Models always produce the same output when setting top_k to 0, still using min_p

### ISSUE THREAD

**Issue by m-from-space** [abetlen/llama-cpp-python#1154]
So I just installed the latest release and immediately realized that it always produces the same token sequence (meaning the same answer for the same question etc.). Tested multiple models, all the same. **Switching back to 0.2.37 fixes the issue.** Reinstalling 0.2.38 makes the issue reappear.

Did some default parameter or sampling behavior change?

As usual I use the following parameters (they worked flawlessly before):

```
temperature 1.19
repeat_penalty 1.0
min_p 0.02
top_p 1.0 (disabled)
top-k 0 (disabled)
```

System: Ubuntu 20.04, CUDA

**Comment by Ph0rk0z** [comment]
Did you check the seed?

**Comment by m-from-space** [comment]
> Did you check the seed?

Thanks for asking. For the seed I use `-1`. The issue only occurs in version 0.2.38, not earlier.

Did the seed behavior change? In the API reference it says, that default seed is `None`. Since when is that?

**Comment by m-from-space** [comment]
Alright, I found the devil. It's about the `top_k` parameter. Something must have changed. Setting it to 0 to disable it, seems to be an issue all of a sudden.

Would be nice if somebody here could check if it's actually about `llama-cpp-python` or `llama.cpp`.

**Comment by abetlen** [comment]
Hey @m-from-space could be a bug, here is the sample method https://github.com/abetlen/llama-cpp-python/blob/310fbf4e494e30d6a78be1b8a6b9be81f61204e7/llama_cpp/_internals.py#L689 it should be nearly identical to the one in llama.cpp

**Comment by JohannesGaessler** [comment]
I think I can reproduce the issue in llama.cpp and have made a PR for a fix: https://github.com/ggerganov/llama.cpp/pull/5388 .
You are calling `llama_sample_top_k` directly instead of via `sampler_queue`, correct?

**Comment by m-from-space** [comment]
Thank you @JohannesGaessler, I hope this PR will soon make it into `llama-cpp-python`. Version `0.2.38` and `0.2.39` still have this issue. So I am leaving this open for now.

**Comment by abetlen** [comment]
@JohannesGaessler thanks for the fix!

@m-from-space yup should have a new release tonight or early tomorrow!

**Comment by m-from-space** [comment]
This is fixed for me in [0.2.41], probably [0.2.40] as well, didn't test it.
